Topic/refactor server clean

telephon opened the pull request:
Cleans up the Server implementation, which needed it.

This PR is still being developed, but it shouldn't break and can be tested – it is planned for ````3.8````.

This fixes #774 and fixes #889: instead of just assuming that the server is dead, it is taken to be unresponsive. It also fixes #1484.

It factors out all statusWatcher and recording functionality to a separate class (````ServerStatusWatcher```` and ````Recorder````).  This branch also implements the updating of recording display in the IDE.

There is some more refactoring, in the server code itself, the Volume class, and also the allocation of buffers in Buffer. It also moves the recording timer to the synthdef, to make the implementation more reliable eventually and also easier for other editors to implement.

While I took care to keep the old interface, this refactor **may break code** that relies on instance variables in ````Server````, these problems are usually easy to solve, by writing, e.g. ````this.avgCPU```` instead of ````avgCPU````.

timblechmann commented on editors/sc-ide/widgets/main_window.cpp:
```diff
@@ -658,6 +658,8 @@ void MainWindow::createMenus()
     menu->addAction( mMain->scServer()->action(ScServer::PlotTree) );
     menu->addAction( mMain->scServer()->action(ScServer::DumpOSC) );
+	menu->addAction( mMain->scServer()->action(ScServer::Record) );
+	menu->addAction( mMain->scServer()->action(ScServer::PauseRecord) );
```
indentation

timblechmann commented on editors/sc-ide/widgets/audio_status_box.cpp:
```diff
@@ -54,6 +54,7 @@ AudioStatusBox::AudioStatusBox(ScServer *server, QWidget *parent):
     setLayout(layout);
 
     server->action(ScServer::Record)->setProperty("keep_menu_open", true);
+	server->action(ScServer::PauseRecord)->setProperty("keep_menu_open", true);
```
indentation

timblechmann commented on editors/sc-ide/core/sc_server.cpp:
```diff
@@ -314,56 +319,66 @@ void ScServer::sendVolume( float volume )
+        if(mIsRecordingPaused != flag) {
+            mIsRecordingPaused = flag;
+            if(flag) {
+                mLang->evaluateCode( QString("ScIDE.defaultServer.pauseRecording"), true );
```
QStringLiteral

timblechmann commented on editors/sc-ide/core/sc_server.cpp:
```diff
@@ -314,56 +319,66 @@ void ScServer::sendVolume( float volume )
-    else
-        return seconds(0);
+    static const QString startRecordingCommand("ScIDE.defaultServer.record");
+    static const QString stopRecordingCommand("ScIDE.defaultServer.stopRecording");
```
`static const QString` is so qt4 ... we have qt5 and `QStringLiteral`, which is more efficient

timblechmann commented on editors/sc-ide/core/sc_server.hpp:
```diff
@@ -163,9 +170,9 @@ private slots:
     VolumeWidget *mVolumeWidget;
-    QTimer mRecordTimer;
-    boost::chrono::system_clock::time_point mRecordTime;
+    int mRecordTime;
```
what is `mRecordTime`? it used to be a timepoint, now it is an int ... of what? apples? seconds? days?

telephon replied:
I know, boost supplies types for that. I had many troubles with those types, so I just decided: it is just a display. We could call it ````mRecordSeconds```` for clarity?

telephon replied:
Done in 36de7e1eb7769d498bf9756a9040d839e9a96170

timblechmann commented on editors/sc-ide/core/sc_server.cpp:
```diff
@@ -378,8 +393,12 @@ void ScServer::onScLangReponse( const QString & selector, const QString & data )
+    static QString startRecordingSelector("recordingStarted");
+    static QString pauseRecordingSelector("recordingPaused");
+    static QString stopRecordingSelector("recordingStopped");
+    static QString recordingDurationSelector("recordingDuration");
```
at one point, this should probably all be converted to `QStringLiteral`

timblechmann replied:
at one point, this should probably all be converted to `QStringLiteral`

telephon replied:
OK, I'll remember to do that - thank you!

telephon replied:
Just did this, I got this:
````
supercollider/editors/sc-ide/core/sc_server.cpp:63:11: error: unknown type name 'QStringLiteral'; did
      you mean 'QLatin1Literal'?
    const QStringLiteral synthServerCategory(tr("Sound Synthesis Server"));
          ^~~~~~~~~~~~~~
          QLatin1Literal
````

timblechmann replied:
http://woboq.com/blog/qstringliteral.html
http://doc.qt.io/qt-5/qstring.html#QStringLiteral
(google hits one and two)

`QString xxx = QStringLiteral("yyy")`

telephon replied:
I had seen that. I don't see why I get the error though.

````const QStringLiteral synthServerCategory(tr("Sound Synthesis Server"));````

does nothing of the form ````QString xxx = QStringLiteral("yyy")````.

At least I don't recognize that form in it.

timblechmann replied:
`QStringLiteral` is a macro, not a type. it is a macro to efficiently create `QString`s without going through the `QString(const char *)` constructor at run-time

telephon replied:
So, after some meditation on your reply I realized: you mean it *should* be in the form ````QString xxx = QStringLiteral("yyy")````, not that this is the error.

timblechmann commented on editors/sc-ide/core/sc_server.cpp:
```diff
@@ -392,9 +411,29 @@ void ScServer::onScLangReponse( const QString & selector, const QString & data )
+    }
+    else if (selector == stopRecordingSelector) {
+        setRecording(false);
+    }
```
it might be reasonable to introduce a `Q_PROPERTY` (http://doc.qt.io/qt-5/properties.html) with an `enum RecordingState { stopped, running, paused}` and propagate all changes via a `recordingStateChanged` notification (http://doc.qt.io/qt-5/signalsandslots.html#signals-and-slots).

telephon replied:
Yes, indeed I currently have the problem that for some reason, that code breaks. I'm a bit at loss why this is, 
````
s.boot;

s.mute; // see the red “m”
s.unmute;

s.record; // this sends “recordingStarted” to the IDE
s.stopRecording; // this sends “recordingStopped”

s.mute; // now this doesn’t display any more
s.unmute;
````

This even happens when I comment out the above code. I checked carefully and it is nothing in sclang.

telephon replied:
OK, I think I found it.

telephon replied:
no, persists, this is hidden in an interaction of ````SimpleController```` with its own calls.

telephon replied:
Nope, was something else: see #1657

timblechmann commented on editors/sc-ide/core/sc_server.cpp:
```diff
@@ -314,56 +319,66 @@ void ScServer::sendVolume( float volume )
+                mLang->evaluateCode( QString("ScIDE.defaultServer.pauseRecording"), true );
+            } else {
+                if(mIsRecording) {
+                setRecording(true);
```
indentation

telephon replied:
OK, I'll fix them all.

timblechmann commented:
have only had a quick look into the ide-part of this PR:
* please use the indentation of the surrounding code
* the more complex the handling of state becomes, the more fragile imperative code will be ... things become more robust when thinking more in terms of 'state' and 'change notification', e.g. by using `Q_PROPERTY` ...

telephon commented:
>the more complex the handling of state becomes, the more fragile imperative code will be ... things become more robust when thinking more in terms of 'state' and 'change notification', e.g. by using Q_PROPERTY ...

Looks reasonable. Currently, I barely get around just extending code that is there already. Also I debug with make in the terminal, and this gives me very few useful cues.

So if you could help out with this one a little it would be really good.

telephon commented:
I've changed the milestone for this – the reason is that I don't really see how we can deal with a separate 3.8 branch currently. There is too much work involved in keeping it up to date. So my suggestion is to merge this one for 3.7.1

Opinions?

crucialfelix commented:
I suggest we release 3.7.0 :)



On Mon, Nov 9, 2015 at 7:08 PM Julian Rohrhuber <notifications@github.com>
wrote:

> I've changed the milestone for this – the reason is that I don't really
> see how we can deal with a separate 3.8 branch currently. There is too much
> work involved in keeping it up to date. So my suggestion is to merge this
> one for 3.7.1
>
> Opinions?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/supercollider/supercollider/pull/1655#issuecomment-155142936>
> .
>

telephon commented:
Yes, please go ahead. Should I add any known issues?

telephon commented:
As we now have a 3.7 release branch, I'd like to merge this finally, before it gets stale. Any objections?

telephon commented:
Ok, if there are no objections, I'll merge this into ```master```.

timblechmann commented:
in september i had two review comments:
* please use the indentation of the surrounding code
* the more complex the handling of state becomes, the more fragile imperative code will be ... things become more robust when thinking more in terms of 'state' and 'change notification', e.g. by using `Q_PROPERTY` ...

also i had suggested to use QStringLiteral to construct QStrings, which apparently never made it into this PR.

architecture-wise, i'd **strongly** suggest to look the second point, as the code as it is becomes much more robust to think more in terms of "state" and "state transitions", as the current imperative code will at one point end up being unmaintainable.

also:
* you mentioned that it *may break code*. this should probably be noted in the user-visible changelog.
* are there conflicts with #1715?

----

side note: i'd strongly suggest to avoid large pull requests, because they are often too much code to review. working in small incremental steps has the advantage that the code gets reviewed better, and both developer and review can follow what is going on.

telephon commented:
Thank you for the comments. 

Re (1): ok, I'll check again. Probably my editor settings were changed unnoticed.
Re (2): this is something I might be able to learn over a year's time, which is on a time scale that is beyond the time we should wait for the merge. Given the size of the changes, it is probably advisable to do this in a separate step.

I can only repeat: yes, whenever possible it is advisable to split into smaller parts. In the current case, it isn't afaics.

>are there conflicts with #1715?

Probably yes, but maybe not.

telephon commented:
>also i had suggested to use QStringLiteral to construct QStrings, which apparently never made it into this PR.

I am sure I tried, but probably I ran into difficulties somewhere. I'd be really more happy to resolve such things once this is merged. The branch has been working fine here since September.

In general, I am probably not the best person to try and refactor the IDE code, since I am an c++ amateur.

timblechmann commented:
well, bogosort is also "working" ...

everyone starts as amateur, but one can try to learn e.g. by address review comments. personally i learned a **lot** when other people reviewed my code and taking criticism personal is absolutely counterproductive.
there are two reasons why i'm still actively reviewing code for supercollider: i'd like to keep it stable enough and prevent architectural design faults (as i need a reliable tool for my own works) and to guide other people through the codebase to make sure that people get an understanding about it.

"please merge now, i'll address the review comments later" ... well, usually the workflow is the other way round ...

fwiw, i gave suggestions how my review comments could be addressed. questions are: who is going do the cleanups? or will the bits just start to rot? as for #1715, i guess it is me who will have to rework it?

telephon commented:
Tim, your suggestions are very welcome, and usually I spend several hours trying to follow them. It is only when I realise that it is further beyond what I can do, when I suggest that someone else (not necessarily you) might have to take a look.

>fwiw, i gave suggestions how my review comments could be addressed. questions are: who is going do the cleanups? 

Well, some of your suggestions were indeed cleanups (I'll take care of the tabs/spaces). Others were architectural enhancements, which I can't implement. It would be nice if we could cooperate on issues that go beyond the expertise of a single person.

>or will the bits just start to rot? as for #1715, i guess it is me who will have to rework it?

Unfortunately, it looks like the changes do conflict, it seems that you hadn't trusted my refactoring enough to make yours on top of it. I tried (the whole evening) to get git to work to combine them in my own private repo, and simplify the current branch as you suggested, but gave up. So the only way I can see is that one of us merges first, and then the other one has to do some adjustments.

telephon commented:
>well, bogosort is also "working" ...

all I meant that the sclang side refactoring was successful not only from a systematic point of view, but also empirically. If the IDE needs some refactoring , this seems like a separate thing

timblechmann commented:
>     well, bogosort is also "working" ...
> 
> all I meant that the sclang side refactoring was successful not only
> from a systematic point of view, but also empirically. If the IDE needs
> some refactoring , this seems like a separate thing

there is the notion of 'constant refactoring'. seeing something that
should be done differently, you try to improve it. has the advantage
that one can learn about different areas.

in this case, there is a pre-qt5 coding style and i had suggested to use
a post-qt5 coding style.

whatever, forget about it

timblechmann commented:
>     or will the bits just start to rot? as for #1715
>     <https://github.com/supercollider/supercollider/pull/1715>, i guess
>     it is me who will have to rework it?
> 
> Unfortunately, it looks like the changes do conflict, it seems that you
> hadn't trusted my refactoring enough to make yours on top of it. I tried
> (the whole evening) to get git to work to combine them in my own private
> repo, and simplify the current branch as you suggested, but gave up. So
> the only way I can see is that one of us merges first, and then the
> other one has to do some adjustments.

forget about it, i'll do it. fwiw, trying to merge changes in small
blocks, things can be reviewed and merged faster, so that there will be
less conflicts with other people's work.

telephon commented:
ok, I'll keep that in mind - thanks for your patience.

telephon commented:
(btw. because you mentioned it: I did actually change QString -> QStringLiteral, in ef85fbe)

telephon commented:
@timblechmann you wrote:
>please use the indentation of the surrounding code

I have checked all the files, but couldn't find those you refer to. Any hint? Thanks!

telephon commented:
I think this would be good to merge, and then #1715 on top of it. Does anyone else want to hit the dangerous green button?

(If it would be better to lose the commit history of this branch, and just have the changes sorted according to files, I could try and do this).

telephon commented:
ping

telephon commented:
@crucialfelix and @timblechmann:  I'd merge to master, and then we look after #1715 (I'll do what I can)?

telephon commented:
ok, I'll look after it once more. But then we should really merge if there is no substantial argument against it.

crucialfelix commented:
I can fix the conflicts and merge. Its a bit tricky to do.

telephon commented:
That would be good, I've just spent an hour or so and failed. It is really just one file (`Server.sc`), in the code easy to resolve, but I somehow keep getting the git workflow wrong.

crucialfelix commented:
It is annoying to do. I always use FileMerge though there are other better
tools I suppose.

On Thu, Mar 17, 2016 at 9:15 PM Julian Rohrhuber <notifications@github.com>
wrote:

> That would be good, I've just spent an hour or so and failed. It is really
> just one file (Server.sc), in the code easy to resolve, but I somehow
> keep getting the git workflow wrong.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/supercollider/supercollider/pull/1655#issuecomment-198064704>
>
-- 
https://twitter.com/crucialfelix
https://medium.com/@crucialfelix
http://www.mattermind.com/

telephon commented:
My main trouble is not the conflict (which seemed easy to manually merge), but the procedure to (1) get the correct commits and (2) not mess up the submodules.

crucialfelix commented:
In future we will just have to reject any PR that are this big. It needs to concentrate on single features without affecting zillions of files and other functions.

telephon commented:
@crucialfelix I've taken another look at this. I still do not understand how such a large refactoring could work with git. It seems orthogonal, just too many things that need to be moved around.

I do see that there are some separate issues in here, but the general picture stays the same.

telephon commented:
@crucialfelix: I can try to merge this in a new branch and get rid of all the conflicts and individual commits. But this is tedious – does it make sense at all?

telephon commented:
OK, here is a branch without merge conflicts. The only relevant commits are the ones after `relolving merge conflicts`. 

Here you can see them:
https://github.com/telephon/supercollider/tree/topic-server-refactoring-post-3.8

When I make a PR, I only get these, added with all the old ones on top, even though I had squashed them with a rebase. Maybe something went wrong there.

in any case, this is the text of the PR:

This is a rather large PR, but is now separated into commits of functional groupings. The only larger chunk is in the last one, this would be very hard to separate. This PR fixes the conflicts in #1655. It can serve as a basis for cherry-picking, so no need to merge it all in one go.
